ERCS/erc-7920.md

```diff
@@ -0,0 +1,299 @@
+
+To verify that an individual message `mₓ` was included in a composite signature:
+
+1. Verify the signature on the `merkleRoot`:
+
+   ```
+   recoveredSigner = ecrecover(merkleRoot, signature)
+   isValidSignature = (recoveredSigner == expectedSigner)
+   ```
+
+2. Compute the leaf node for message `mₓ` and verify it's path to the Merkle root, using the proof:
+   ```
+   leaf = keccak256(encode(mₓ))
+   isValidProof = _verifyMerkleProof(leaf, merkleProof, merkleRoot)
+   ```
+
+The message is verified if and only if (1) and (2) succeed.
+
+```
+isVerified = isValidSignature && isValidProof
+```
+
+#### Solidity Example
+
+Snippet below verifies a composite signature on-chain. The entrypoint, `placeOrder()` is called by a paymaster to sponsor an operation. It verifies the composite signature using the process outlined above.
+
+```solidity
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+error Unauthorized();
+
+contract ExampleVerifier {
+    bytes32 public immutable DOMAIN_SEPARATOR;
+    bytes32 private constant MESSAGE_TYPEHASH =
+        keccak256("PlaceOrder(bytes32 orderId, address user)");
+
+    constructor() {
+        DOMAIN_SEPARATOR = keccak256(
+            abi.encode(
+                keccak256(
+                    "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"
+                ),
+                keccak256(bytes("MyApp")),
+                keccak256(bytes("1.0.0")),
+                block.chainid,
+                address(this)
+            )
+        );
+    }
+
+    function placeOrder(
+        bytes32 orderId,
+        address user,
+        bytes calldata signature,
+        bytes32 merkleRoot,
+        bytes32[] calldata proof
+    ) public {
+        bytes32 message = keccak256(
+            abi.encodePacked(
+                "\x19\x01",
+                DOMAIN_SEPARATOR,
+                keccak256(abi.encode(MESSAGE_TYPEHASH, orderId, user))
+            )
+        );
+
+        if (
+            !_verifyCompositeSignature(
+                message,
+                proof,
+                merkleRoot,
+                signature,
+                user
+            )
+        ) {
+            revert Unauthorized();
+        }
+
+        // DO STUFF
+    }
+
+    function _verifyCompositeSignature(
+        bytes32 message,
+        bytes32[] calldata proof,
+        bytes32 merkleRoot,
+        bytes calldata signature,
+        address expectedSigner
+    ) internal view returns (bool) {
+        if (!_verifyMerkleProof(message, proof, merkleRoot)) {
+            return false;
+        }
+
+        return recover(merkleRoot, signature) == expectedSigner;
+    }
+
+    function _verifyMerkleProof(
+        bytes32 leaf,
+        bytes32[] calldata proof,
+        bytes32 root
+    ) internal pure returns (bool) {
+        bytes32 computedRoot = leaf;
+        for (uint256 i = 0; i < proof.length; ++i) {
+            if (computedRoot < proof[i]) {
```

Comment:
Specification should reflect ordering requirements - unless you intentionally don't want to. 

Reference Implementation (Lines 160-168):
// Enforces ordering during verification:
if (computedRoot < proof[i]) {
    computedRoot = keccak256(abi.encodePacked(computedRoot, proof[i]));
} else {
    computedRoot = keccak256(abi.encodePacked(proof[i], computedRoot));
}

Suggested Spec Addition:
"Leaves MUST be ordered lexicographically before tree construction"